module/zfs/zfs_vnops.c

```diff
@@ -1094,6 +1094,17 @@ zfs_clone_range(znode_t *inzp, uint64_t *inoffp, znode_t *outzp,
 
 	ASSERT(!outzfsvfs->z_replay);
 
+	/*
+	 * If source and destination are not the same filesystem
+	 * neither may be encrypted.
+	 * TODO: Support of cloning encrypted datasets with identical keys
+	 */
+	if (inzfsvfs != outzfsvfs &&
+	    (inos->os_encrypted || outos->os_encrypted)) {
```

Comment:
Couldn't we just do here:
```suggestion
	if (inos->os_encrypted != outos->os_encrypted) {
```
We already have `BP_IS_PROTECTED(&bps[0])` and cross-dataset check below.

Reply:
Do you think all the following comparsions and dmu_read_l0_bps() are needed when we know in advance that we bail out? Maybe removing the second check with BP_IS_PROTECTED(&bps[0]) instead?

Reply:
The BP_IS_PROTECTED(&bps[0]) checks after reading the BP that if the source BP is encrypted and we are not on the same filesystem - do not block clone. What we did not check anywhere is that if the source is not encrypted and the destination is - this currently leads in master and in 2.2.0 to copying unencrypted blocks into an encrypted dataset. But IMO when we know that the source dataset (or the target dataset) are encrypted and not the same we don't have to continue. If I apply the change you suggested, then BP_IS_PROTECTED(&bps[0]) will fire only in encrypted -> encrypted case, which I am fine with, so I will update the patch.

Reply:
Patch updated.